apm-agent-core/src/main/java/co/elastic/apm/agent/impl/sampling/ProbabilitySampler.java

```diff
@@ -54,10 +55,14 @@
     private final long higherBound;
     private final double sampleRate;
 
+    // Because header value only contains sampling rate, we can cache it here
+    private final String traceStateHeader;
+
     private ProbabilitySampler(double samplingRate) {
         higherBound = (long) (Long.MAX_VALUE * samplingRate);
         lowerBound = -higherBound;
         this.sampleRate = samplingRate;
+        traceStateHeader = TraceState.buildHeaderString(samplingRate);
```

Comment:
I don't think we need this dependency in `TraceState` - enough to be able to get `sampleRateAsString` here and use the the `StringBuilder` in `TraceState` to append to the header when required. This means both no allocation and no dependency.
Sorry for nagging, but conceptually - a sampler shouldn't be aware of `tracestate`